File: parsers/kotlin/src/main/kotlin/verity/passport/prover/Poseidon2.kt
package verity.passport.prover

import java.math.BigInteger

/**
 * Poseidon2 hash function implementation over the BN254 scalar field.
 *
 * Implements the sponge construction with a 4-element state (rate=3, capacity=1),
 * using the round constants and internal matrix diagonal defined in [Poseidon2Constants].
 * All arithmetic is performed modulo [P] (the BN254 scalar field prime).
 */
public object Poseidon2 {

    /** BN254 scalar field prime. */
    public val P: BigInteger = BigInteger("21888242871839275222246405745257275088548364400416034343698204186575808495617")

    private val ZERO = BigInteger.ZERO
    private val RATE = 3
    private val TWO_POW_64 = BigInteger.ONE.shiftLeft(64)

    private fun add(a: BigInteger, b: BigInteger): BigInteger = (a + b).mod(P)
    private fun mul(a: BigInteger, b: BigInteger): BigInteger = (a * b).mod(P)

    private fun sBox(x: BigInteger): BigInteger {
        val s = mul(x, x)
        return mul(mul(s, s), x)
    }

    private fun matMul4x4(state: Array<BigInteger>) {
        val t0 = add(state[0], state[1])
        val t1 = add(state[2], state[3])
        var t2 = add(state[1], state[1])
        t2 = add(t2, t1)
        var t3 = add(state[3], state[3])
        t3 = add(t3, t0)
        var t4 = add(t1, t1)
        t4 = add(t4, t4)
        t4 = add(t4, t3)
        var t5 = add(t0, t0)
        t5 = add(t5, t5)
        t5 = add(t5, t2)
        val t6 = add(t3, t5)
        val t7 = add(t2, t4)
        state[0] = t6
        state[1] = t5
        state[2] = t7
        state[3] = t4
    }

    private fun internalMatMul(state: Array<BigInteger>, diag: Array<BigInteger>) {
        var sum = ZERO
        for (s in state) sum = add(sum, s)
        for (i in state.indices) {
            state[i] = add(mul(state[i], diag[i]), sum)
        }
    }

    /**
     * Applies the full Poseidon2 permutation to a 4-element state.
     *
     * Consists of [rfFirst] full rounds, [rp] partial rounds, and [rfFirst] more full rounds,
     * using the round constants from [Poseidon2Constants.ROUND_CONSTANTS] and the
     * internal matrix diagonal from [Poseidon2Constants.INTERNAL_MATRIX_DIAGONAL].
     *
     * @param inputs 4-element array of field elements to permute.
     * @return A new 4-element array containing the permuted state.
     */
    public fun permutation(inputs: Array<BigInteger>): Array<BigInteger> {
        val rfFirst = 4
        val rp = 56
        val pEnd = rfFirst + rp
        val numRounds = rfFirst + rp + rfFirst

        val state = inputs.copyOf()
        val rc = Poseidon2Constants.ROUND_CONSTANTS
        val diag = Poseidon2Constants.INTERNAL_MATRIX_DIAGONAL

        matMul4x4(state)

        for (r in 0 until rfFirst) {
            for (i in 0..3) state[i] = add(state[i], rc[r][i])
            for (i in 0..3) state[i] = sBox(state[i])
            matMul4x4(state)
        }

        for (r in rfFirst until pEnd) {
            state[0] = add(state[0], rc[r][0])
            state[0] = sBox(state[0])
            internalMatMul(state, diag)
        }

        for (r in pEnd until numRounds) {
            for (i in 0..3) state[i] = add(state[i], rc[r][i])
            for (i in 0..3) state[i] = sBox(state[i])
            matMul4x4(state)
        }

        return state
    }

    /**
     * Hashes a list of field elements using the Poseidon2 sponge construction.
     *
     * The capacity element is initialised as `len(inputs) * 2^64` to provide
     * domain separation. Inputs are absorbed [RATE] elements at a time; the
     * permutation is applied whenever the internal cache is full. The first
     * element of the final permuted state is returned as the digest.
     *
     * @param inputs List of BN254 field elements to hash.
     * @return The Poseidon2 hash as a [BigInteger] field element.
     */
    public fun hash(inputs: List<BigInteger>): BigInteger {
        val iv = mul(BigInteger.valueOf(inputs.size.toLong()), TWO_POW_64)

        val state = arrayOf(ZERO, ZERO, ZERO, iv)
        val cache = arrayOf(ZERO, ZERO, ZERO)
        var cacheSize = 0

        for (input in inputs) {
            if (cacheSize == RATE) {
                for (i in 0 until RATE) state[i] = add(state[i], cache[i])
                cache[0] = ZERO; cache[1] = ZERO; cache[2] = ZERO
                cacheSize = 0
                val perm = permutation(state)
                state[0] = perm[0]; state[1] = perm[1]; state[2] = perm[2]; state[3] = perm[3]
            }
            cache[cacheSize] = input
            cacheSize++
        }

        for (i in 0 until cacheSize) state[i] = add(state[i], cache[i])
        val result = permutation(state)

        return result[0]
    }

    /**
     * Hashes one or more hex-encoded field elements using the Poseidon2 sponge construction.
     *
     * Each string is decoded via [hexToField] before hashing.
     *
     * @param hexInputs Hex-encoded field element strings (with or without `0x` prefix).
     * @return The Poseidon2 hash as a [BigInteger] field element.
     */
    public fun hash(vararg hexInputs: String): BigInteger {
        return hash(hexInputs.map { hexToField(it) })
    }

    /**
     * Decodes a hex string into a BN254 field element, reducing modulo [P].
     *
     * @param hex Hex string with optional `0x` prefix.
     * @return The field element as a [BigInteger] in `[0, P)`.
     */
    public fun hexToField(hex: String): BigInteger {
        val s = if (hex.startsWith("0x")) hex.substring(2) else hex
        return BigInteger(s, 16).mod(P)
    }

    /**
     * Encodes a BN254 field element as a zero-padded 32-byte hex string with `0x` prefix.
     *
     * @param fe Field element to encode; reduced modulo [P] before encoding.
     * @return Hex string of the form `"0x" + 64 hex digits`.
     */
    public fun fieldToHex(fe: BigInteger): String {
        return "0x" + fe.mod(P).toString(16).padStart(64, '0')
    }

    /**
     * Packs a byte array into a list of BN254 field elements in little-endian field order.
     *
     * Bytes are grouped into chunks of up to [bytesPerField] bytes each. The first chunk
     * may be smaller than [bytesPerField] to consume the leading bytes. Each chunk is
     * interpreted as a big-endian unsigned integer and reduced modulo [P]. The resulting
     * list is reversed so that the chunk containing the lowest-address bytes appears last,
     * matching the Noir circuit packing convention.
     *
     * @param bytes Raw bytes to pack.
     * @param bytesPerField Maximum bytes per field element (default 31, safe for BN254).
     * @return List of field elements in little-endian field order.
     */
    public fun packBytesIntoFields(bytes: ByteArray, bytesPerField: Int = 31): List<BigInteger> {
        val numFields = (bytes.size + bytesPerField - 1) / bytesPerField
        val firstFieldSize = bytes.size - (numFields - 1) * bytesPerField
        var offset = 0

        return buildList(numFields) {
            var value = ZERO
            for (i in 0 until firstFieldSize) {
                value = value.shiftLeft(8).add(BigInteger.valueOf((bytes[offset++].toInt() and 0xFF).toLong()))
            }
            add(value.mod(P))

            for (f in 1 until numFields) {
                value = ZERO
                for (i in 0 until bytesPerField) {
                    value = value.shiftLeft(8).add(BigInteger.valueOf((bytes[offset++].toInt() and 0xFF).toLong()))
                }
                add(value.mod(P))
            }
        }.reversed()
    }
}

File: parsers/swift/Sources/PassportProverLib/Poseidon2.swift
import Foundation
import BigInt

/// Poseidon2 sponge hash over BN254 scalar field.
/// Ported directly from the Kotlin implementation.
public enum Poseidon2 {

    public static let P = BigUInt("21888242871839275222246405745257275088548364400416034343698204186575808495617")

    private static let ZERO = BigUInt(0)
    private static let RATE = 3
    private static let TWO_POW_64 = BigUInt(1) << 64

    private static func add(_ a: BigUInt, _ b: BigUInt) -> BigUInt {
        (a + b) % P
    }

    private static func mul(_ a: BigUInt, _ b: BigUInt) -> BigUInt {
        (a * b) % P
    }

    private static func sBox(_ x: BigUInt) -> BigUInt {
        let s = mul(x, x)
        return mul(mul(s, s), x)
    }

    private static func matMul4x4(_ state: inout [BigUInt]) {
        let t0 = add(state[0], state[1])
        let t1 = add(state[2], state[3])
        var t2 = add(state[1], state[1])
        t2 = add(t2, t1)
        var t3 = add(state[3], state[3])
        t3 = add(t3, t0)
        var t4 = add(t1, t1)
        t4 = add(t4, t4)
        t4 = add(t4, t3)
        var t5 = add(t0, t0)
        t5 = add(t5, t5)
        t5 = add(t5, t2)
        let t6 = add(t3, t5)
        let t7 = add(t2, t4)
        state[0] = t6
        state[1] = t5
        state[2] = t7
        state[3] = t4
    }

    private static func internalMatMul(_ state: inout [BigUInt], _ diag: [BigUInt]) {
        var sum = ZERO
        for s in state { sum = add(sum, s) }
        for i in state.indices {
            state[i] = add(mul(state[i], diag[i]), sum)
        }
    }

    public static func permutation(_ inputs: [BigUInt]) -> [BigUInt] {
        let rfFirst = 4
        let rp = 56
        let pEnd = rfFirst + rp
        let numRounds = rfFirst + rp + rfFirst

        var state = inputs
        let rc = Poseidon2Constants.roundConstants
        let diag = Poseidon2Constants.internalMatrixDiagonal

        matMul4x4(&state)

        for r in 0 ..< rfFirst {
            for i in 0 ... 3 { state[i] = add(state[i], rc[r][i]) }
            for i in 0 ... 3 { state[i] = sBox(state[i]) }
            matMul4x4(&state)
        }

        for r in rfFirst ..< pEnd {
            state[0] = add(state[0], rc[r][0])
            state[0] = sBox(state[0])
            internalMatMul(&state, diag)
        }

        for r in pEnd ..< numRounds {
            for i in 0 ... 3 { state[i] = add(state[i], rc[r][i]) }
            for i in 0 ... 3 { state[i] = sBox(state[i]) }
            matMul4x4(&state)
        }

        return state
    }

    public static func hash(_ inputs: [BigUInt]) -> BigUInt {
        let iv = mul(BigUInt(inputs.count) , TWO_POW_64)

        var state: [BigUInt] = [ZERO, ZERO, ZERO, iv]
        var cache: [BigUInt] = [ZERO, ZERO, ZERO]
        var cacheSize = 0

        for input in inputs {
            if cacheSize == RATE {
                for i in 0 ..< RATE { state[i] = add(state[i], cache[i]) }
                cache = [ZERO, ZERO, ZERO]
                cacheSize = 0
                state = permutation(state)
            }
            cache[cacheSize] = input
            cacheSize += 1
        }

        for i in 0 ..< cacheSize { state[i] = add(state[i], cache[i]) }
        let result = permutation(state)

        return result[0]
    }

    public static func hash(_ hexInputs: String...) -> BigUInt {
        hash(hexInputs.map { hexToField($0) })
    }

    public static func hexToField(_ hex: String) -> BigUInt {
        let s = hex.hasPrefix("0x") ? String(hex.dropFirst(2)) : hex
        return BigUInt(s, radix: 16)! % P
    }

    public static func fieldToHex(_ fe: BigUInt) -> String {
        let reduced = fe % P
        let hex = String(reduced, radix: 16)
        return "0x" + String(repeating: "0", count: max(0, 64 - hex.count)) + hex
    }

    public static func packBytesIntoFields(_ bytes: Data, bytesPerField: Int = 31) -> [BigUInt] {
        let numFields = (bytes.count + bytesPerField - 1) / bytesPerField
        var fields: [BigUInt] = []

        let firstFieldSize = bytes.count - (numFields - 1) * bytesPerField
        var offset = 0

        var value = ZERO
        for _ in 0 ..< firstFieldSize {
            value = (value << 8) + BigUInt(bytes[bytes.startIndex + offset])
            offset += 1
        }
        fields.append(value % P)

        for _ in 1 ..< numFields {
            value = ZERO
            for _ in 0 ..< bytesPerField {
                value = (value << 8) + BigUInt(bytes[bytes.startIndex + offset])
                offset += 1
            }
            fields.append(value % P)
        }

        return fields.reversed()
    }

    /// Convenience: pack [UInt8] array.
    public static func packBytesIntoFields(_ bytes: [UInt8], bytesPerField: Int = 31) -> [BigUInt] {
        packBytesIntoFields(Data(bytes), bytesPerField: bytesPerField)
    }
}
